src/charm.py

```diff
@@ -19,6 +24,183 @@ class KratosCharm(CharmBase):
+            self._container.add_layer(self._container_name, new_layer, combine=True)
+            logger.info("Pebble plan updated with new configuration, replanning")
+            self._container.replan()
+
+        # Compare changes in kratos config
+        current_config = self._container.pull(self._config_file_path).read()
+        if current_config != self._config:
+            self._container.push(self._config_file_path, self._config, make_dirs=True)
+            logger.info("Updated kratos config")
+            self._container.replan()
+
+    def _get_database_relation_info(self) -> dict:
+        """Get database info from relation data bag."""
+        relation_id = self.database.relations[0].id
+        relation_data = self.database.fetch_relation_data()[relation_id]
+
+        return {
+            "username": relation_data["username"],
+            "password": relation_data["password"],
+            "endpoints": relation_data["endpoints"],
+        }
+
+    def _set_default_identity_schema(self) -> None:
+        """Push identity schema into kratos container."""
+        try:
+            with open("src/identity.default.schema.json", encoding="utf-8") as schema_file:
+                schema = schema_file.read()
+                self._container.push(self._identity_schema_file_path, schema, make_dirs=True)
+            logger.info("Pushed configs to kratos container")
+        except (ProtocolError, PathError) as e:
+            logger.error(str(e))
+            self.unit.status = BlockedStatus(str(e))
+
+    def _update_container(self, event) -> None:
+        """Update configs, pebble layer and run database migration."""
+        if not self.unit.is_leader():
+            event.defer()
+            logger.info("Does not have leadership. Deferring event.")
+            self.unit.status = WaitingStatus("Waiting for leadership")
+            return
+
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Kratos container. Deferring event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Kratos container")
+            return
+
+        if not self.model.relations["pg-database"]:
+            event.defer()
+            logger.error("Missing required relation with postgresql")
+            self.model.unit.status = BlockedStatus("Missing required relation with postgresql")
+            return
+
+        if not self.database.is_database_created():
+            event.defer()
+            logger.info("Missing database details. Deferring event.")
+            self.unit.status = WaitingStatus("Waiting for database creation")
+            return
+
+        self._set_default_identity_schema()
+
+        try:
+            self._update_layer()
+        except ChangeError as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus("Failed to replan")
+            return
+
+        self._run_sql_migration()
```

Comment:
What happens if it is a breaking migration? Do we want to always auto run the migration? The ory page urges the user to take a backup before running migrations. Should we warn the user? 

This comment is probably out of scope for this PR.

Reply:
Definitely one of my worries too. I read on ory slack awhile back that it is safe to rerun migration. That's why I put this here for now. I have no idea how upgrades will be handled. Now that I am thinking about it, I don't think the set of migration would change since the kratos cli is tied to the image version. I am going to look into this further soon and tackle that in a separate PR.